Added direct ingestion/delta metrics support

vikramraman opened the pull request:
MONIT-7066: Added direct ingestion/delta metrics support for Java.

Note: @basilisk487 's PR (https://github.com/wavefrontHQ/java/pull/270) includes changes to the WavefrontReporter. I can merge his changes into this PR once it is merged into master before this is merged.

vikramraman commented on java-lib/src/main/java/com/codahale/metrics/DeltaCounter.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class DeltaCounter extends Counter {
+
+  public static final String DELTA_PREFIX = "\u2206"; // ∆: INCREMENT
```
Thanks will refactor to have it in a single place.

vikramraman commented on dropwizard-metrics/3.1/src/main/java/com/wavefront/integrations/metrics/WavefrontReporter.java:
```diff
@@ -463,7 +485,14 @@ private void reportHistogram(String name, Histogram histogram) throws IOExceptio
+      long count = counter.getCount();
+      name = DeltaCounter.DELTA_PREFIX + prefixAndSanitize(name.substring(1), "count");
+      wavefront.send(name, count,clock.getTime() / 1000, source, pointTags);
+      counter.dec(count);
```
@sushantdewan123 `AtomicLong` was used internally in YammerMetrics Counter.

DropWizard 3.0.x uses an internal implementation of `LongAdder`: [3.0.x](https://github.com/dropwizard/metrics/blob/3.0-maintenance/metrics-core/src/main/java/com/codahale/metrics/Counter.java)
DropWizard 3.1.x-3.2.x uses an internal `LongAdderAdapter`: [3.2.x](https://github.com/dropwizard/metrics/blob/3.2-maintenance/metrics-core/src/main/java/com/codahale/metrics/Counter.java)
And DropWizard 4.x uses `java.util.concurrent.atomic.LongAdder`: [4.x](https://github.com/dropwizard/metrics/blob/4.0-maintenance/metrics-core/src/main/java/com/codahale/metrics/Counter.java)

We could account for these variations, but that would complicate things (testing for example). We can override the decrement methods and throw if called and maybe add a `reset()` method that internally calls `super.dec(getCount())`. That way we can guard against the delta counters going negative. Would that work?

sushantdewan123 commented on java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java:
```diff
@@ -0,0 +1,272 @@
+    //TODO: extra validation required?
+    String point = pointToString(name, value, timestamp, source, pointTags);
+    if (point != null) {
+      synchronized(pointsMutex) {
```
This is sort of a bottleneck. But I don't think we are sending high-frequency items via direct ingestion API so bottleneck should not be that bad.
Also, all the points sent to the in-memory buffer via direct ingestion will be lost if the client process is restarted.

panghy replied:
Use a ReadWriteLock and just acquire read locks here and write lock when you flush.

panghy commented on java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java:
```diff
@@ -0,0 +1,272 @@
+    }
+
+    if (timestamp == null) {
+      timestamp = System.currentTimeMillis()/1000;
```
THis should not be necessary, let's assign the time on the server side upon receipt (eliminate clock drift confusion)

panghy commented on java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java:
```diff
@@ -0,0 +1,272 @@
+  static String pointToString(String name, double value, @Nullable Long timestamp, String source,
+                               @Nullable Map<String, String> pointTags) {
+
+    if (name == null || name.isEmpty() || source == null || source.isEmpty()) {
```
StringUtils.isBlankOrEmptyString

panghy commented on java-lib/src/main/java/com/wavefront/api/DataIngestorAPI.java:
```diff
@@ -0,0 +1,33 @@
+ * @author Vikram Raman
+ */
+@Path("/")
+public interface DataIngestorAPI {
```
DataingesterAPI

panghy commented on java-lib/src/main/java/com/wavefront/api/DataIngestorAPI.java:
```diff
@@ -0,0 +1,33 @@
+import javax.ws.rs.core.Response;
+
+/**
+ *
```
javadoc

panghy commented on java-lib/src/main/java/com/wavefront/api/DataIngestorAPI.java:
```diff
@@ -0,0 +1,33 @@
+  @POST
+  @Path("report")
+  @Consumes(MediaType.MULTIPART_FORM_DATA)
+  Response uploadFile(@QueryParam("f") String format,
```
?

panghy commented on java-lib/src/main/java/com/wavefront/api/DataIngestorAPI.java:
```diff
@@ -0,0 +1,33 @@
+  @Path("report")
+  @Consumes({MediaType.APPLICATION_OCTET_STREAM, MediaType.APPLICATION_FORM_URLENCODED,
+      MediaType.TEXT_PLAIN})
+  Response report(@QueryParam("f") String format, InputStream stream) throws IOException;
```
do we need to doc this?

panghy commented on proxy/src/main/java/com/wavefront/agent/QueuedAgentService.java:
```diff
@@ -25,6 +25,7 @@
 import com.wavefront.api.WavefrontAPI;
 import com.wavefront.api.agent.AgentConfiguration;
 import com.wavefront.api.agent.ShellOutputDTO;
+import com.wavefront.common.NamedThreadFactory;
```
unused import?

vikramraman replied:
I've moved `NamedThreadFactory` from `com.wavefront.agent` to `com.wavefront.common`. `NamedThreadFactory` is used in these classes but the import wasn't required earlier as they were under the same package, but is needed now.

panghy commented on proxy/src/main/java/com/wavefront/agent/PushAgent.java:
```diff
@@ -32,6 +32,7 @@
 import com.wavefront.agent.preprocessor.ReportPointTimestampInRangeFilter;
 import com.wavefront.api.agent.AgentConfiguration;
 import com.wavefront.api.agent.Constants;
+import com.wavefront.common.NamedThreadFactory;
```
unused import?

panghy commented on proxy/src/main/java/com/wavefront/agent/PostSourceTagTimedTask.java:
```diff
@@ -3,6 +3,7 @@
 import com.google.common.util.concurrent.RateLimiter;
 
 import com.wavefront.agent.api.ForceQueueEnabledAgentAPI;
+import com.wavefront.common.NamedThreadFactory;
```
unused import?

panghy commented on proxy/src/main/java/com/wavefront/agent/AbstractAgent.java:
```diff
@@ -26,6 +26,7 @@
 import com.wavefront.api.agent.AgentConfiguration;
 import com.wavefront.api.agent.Constants;
 import com.wavefront.common.Clock;
+import com.wavefront.common.NamedThreadFactory;
```
unused import?

panghy commented on proxy/src/main/java/com/wavefront/agent/PostPushDataTimedTask.java:
```diff
@@ -5,6 +5,7 @@
 
 import com.wavefront.agent.api.ForceQueueEnabledAgentAPI;
 import com.wavefront.api.agent.Constants;
+import com.wavefront.common.NamedThreadFactory;
```
unused import?

panghy commented on java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java:
```diff
@@ -0,0 +1,270 @@
+  private Future<?> scheduledFuture;
+
+  //TODO: bounded buffer?
+  private List<String> buffer = new ArrayList<>();
```
There should probably be a limit since a machine with connectivity issue or an invalid token should not collect points forever.

panghy replied:
This needs to be marked volatile

panghy commented on java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java:
```diff
@@ -0,0 +1,270 @@
+
+  private List<String> getPointsBatch() {
+    List<String> points;
+    synchronized(pointsMutex) {
```
formatting

panghy commented on java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java:
```diff
@@ -0,0 +1,270 @@
+      response = directService.report("graphite_v2", is);
+      if (response.getStatusInfo().getFamily() == Response.Status.Family.SERVER_ERROR ||
+          response.getStatusInfo().getFamily() == Response.Status.Family.CLIENT_ERROR) {
+        LOGGER.debug("Error reporting points, respStatus=" + response.getStatus());
```
Probably should look at putting the points back into the buffer (while respecting the maximum size of the buffer)

panghy commented on java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java:
```diff
@@ -0,0 +1,270 @@
+  public void run() {
+    try {
+      this.internalFlush();
+    } catch (Exception ex) {
```
Should catch Throwable

panghy commented on dropwizard-metrics/3.1/src/main/java/com/wavefront/integrations/metrics/WavefrontReporter.java:
```diff
@@ -298,6 +301,12 @@ public WavefrontReporter bindToCloudFoundryService(@NotNull String proxyServiceN
           disabledMetricAttributes);
     }
 
+    public WavefrontReporter buildDirect(String server, String token) {
```
comments for public method?

vikramraman commented:
@panghy Added javadoc for the `buildDirect` method. Also added logic to shutdown the ExecutorService on `close()` (found an issue with client code using the direct SDK not shutting down cleanly because of this).

I've addressed all of @sushantdewan123 comments. Discussed with him offline and we agreed current implementation is the best approach for now.
